iac-extensions/docker/src/main/java/org/sonar/iac/docker/visitors/DockerSymbolVisitor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.visitors;
+
+import org.sonar.iac.common.extension.visitors.InputFileContext;
+import org.sonar.iac.common.extension.visitors.TreeVisitor;
+import org.sonar.iac.docker.symbols.Scope;
+import org.sonar.iac.docker.symbols.Symbol;
+import org.sonar.iac.docker.symbols.Usage;
+import org.sonar.iac.docker.tree.api.ArgInstruction;
+import org.sonar.iac.docker.tree.api.Body;
+import org.sonar.iac.docker.tree.api.DockerImage;
+import org.sonar.iac.docker.tree.api.NewKeyValuePair;
+import org.sonar.iac.docker.tree.api.Variable;
+import org.sonar.iac.docker.utils.ArgumentUtils;
+
+public class DockerSymbolVisitor extends TreeVisitor<InputFileContext> {
+
+  private final Scope globalScope = new Scope();
+  private Scope currentScope = globalScope;
+
+  public DockerSymbolVisitor() {
+    register(Body.class, this::setGlobalScope);
+    register(DockerImage.class, this::setImageScope);
+    register(ArgInstruction.class, this::visitArgInstruction);
+    register(Variable.class, this::visitVariable);
+  }
+
+  public void setGlobalScope(InputFileContext ctx, Body body) {
+    body.setScope(globalScope);
+  }
+
+  public void setImageScope(InputFileContext ctx, DockerImage dockerImage) {
+    currentScope = new Scope(globalScope);
+    dockerImage.setScope(currentScope);
+  }
```

Comment:
We will probably need to store the inheritance information, as by default for an `Image`, inherited global variable are not accessible.
```
ARG global_var=value
FROM scratch
LABEL key=$global_var
# will give : {"key":""}
```
VS
```
ARG global_var=value
FROM scratch
ARG global_var
LABEL key=$global_var
# will give : {"key":"value"}
```

Another use case to consider : global variable are naturally always accessible in FROM instruction :
```
ARG image=scratch
FROM $image
```

Those use cases can probably be processed in separate tickets.

Reply:
Why do you think that they are not accessible? They should be via:
```
currentScope = new Scope(globalScope);
dockerImage.setScope(currentScope);
```

Reply:
As discussed, we need to have a mention of the global variable in a ARG instruction inside the image to have access to it.
Even the [dockerfile reference](https://docs.docker.com/engine/reference/builder/#scope) is not very clear about it as they don't give a full example.

Also, I just saw that this PR does not support ENV at all, shouldn't we ? I didn't saw any reference to this in the Jira ticket.
 (better to have [SONARIAC-549](https://sonarsource.atlassian.net/browse/SONARIAC-549) completed first)
I saw also [this](https://docs.docker.com/engine/reference/builder/#using-arg-variables) in the Dockerfile reference :

> You can use an ARG or an ENV instruction to specify variables that are available to the RUN instruction. Environment variables defined using the ENV instruction always override an ARG instruction of the same name. Consider this Dockerfile with an ENV and ARG instruction.